man/man1/runbody2.1
.TH RUNBODY2 1NEMO "9 February 2019"
.SH NAME
runbody2 \- Ahmad-Cohen N-body code
.SH SYNOPSIS
.nf
\fBrunbody2\fP [parameter=value]
\fBnbody2  \fP\fI < nbody2.in\fP
.fi
.SH DESCRIPTION
\fBrunbody2\fP is a frontend to the NBODY2 fortran program,
running the code in a run directory where all files are stored.
.PP
The new version 2.x interface tries to mimic to run nbody1 in virial units
with a default input and output.
In previous versions the defaults were more geared towards
running the classic interface where data were scaled and times were measured
in crossing times, making comparisons with other codes in NEMO much harder.
This however means it is now more work to make nbody1 run in classic mode.
Parameters names and meaning have not changed, except for the introduction
of kz(19), to work in code (virial) units.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBin=\fP
input file (optional, see \fBnbody=\fP to generate your own input 
data on the fly). If input data are given, it forces KZ(4)=2 (see 
below).  Can also be a directory, in which case the input file for
a restart run is assumed to be the default (fort.4)  in native 
format.
Default: not used.
.TP
\fBoutdir=\fP
Output run directory. This directory should not exist, although
can be overriden with NEMO's standard \fBerror=\fP system keyword.
No default.
.TP
\fBnbody=\fP
Total particle number. The maximum value (NMAX) is compiled into
the NBODY2 program. Note that even if \fIin=\fP is supplied, this
parameter is needed.
No default.
.TP
\fBnfix=\fP
Output frequency of data save or binaries; KZ(3 & 6)
[Default: 1]
.TP
\fBnrand=\fP
Random number sequence skip.
[Default: 0]
.TP
\fBnnbmax=\fP
Maximum number of neighbours. Should be less then LMAX, which  is
compiled into the NBODY2 program.
[Default: 40]
.TP
\fBnrun=\fP
Run identification index.
[Default: 1]
.TP
\fBetai=\fP
Time-step parameter for irregular force polynomial.
[Default: 0.02]
.TP
\fBetar=\fP
Time-step parameter for regular force polynomial.
[Default: 0.04]
.TP
\fBrs0=\fP
Initial radius of neighbour sphere.
[Default: 1]
.TP
\fBdeltat=\fP
Output time interval in units of the crossing.
[Default: 0.25]
.TP
\fBtcrit=\fP
Termination time in units of the crossing time.
[Default: 5]
.TP
\fBqe=\fP
Energy tolerance (restart if DE/E > 5*QE & KZ(2) > 1)
[Default: 2e-5]
.TP
\fBeps=\fP
Potential softening parameter     
[Default: 0.01]
.TP
\fBkz=\fP
Non-zero options for alternative paths (see below) 
[Default: 1,2,1,2,0,0,0,0,0,0,1,0,0,0,0,1,0,0,1,0]
.nf
.ta +0.5i
 1  	COMMON save on unit 1 at end of run (=2: every 100*NMAX steps) [1]
 2  	COMMON save on unit 2 at output (=1); restart if DE/E > 5*QE (=2) [2]
 3  	Basic data written to unit 3 at output time (frequency NFIX) [1]
 4 	Initial conditions on unit 4 (=1: output; =2: input) [2]
 5  	Initial conditions (=0: uniform & isotropic; =1: Plummer  model) [0]
 6  	Output of significant binaries (=2: frequency NFIX) [0]
 7  	Lagrangian radii (=1: unit 6; =2: unit 7; =3: both types) [0]
 8  	Core radius & density centre (N > 20 only) [0]
 9  	Individual bodies printed at output time (MIN(5**KZ(9),N)) [0]
 10  	No unique density centre (skips velocity modification of RS(I)) [0]
 11  	Modification of ETAI & ETAR by tolerance QE [1]
 12  	Inelastic mergers (>1: diagnostic output) [0]
 13  	Escaper removal (R > 2*RTIDE; RTIDE = 10*RSCALE) [0]
 14  	Skip full predictor loop if NNB > KZ(14) = <NNB> & KZ(14) > [0]
 15  	External potential (=1: Plummer model; =2: logarithmic potential) [0]
 16  	No scaling of initial conditions [1]
 17  	Generation of two subsystems (merger experiment) [0]
 18  	Adjustment of coordinates & velocities to c.m. condition [0]
 19  	Use code units for tcrit/deltat [1]
 20  	Not used at present [0]
.fi
.TP
\fBxtpar1=\fP
Mass of external Plummer model (KZ(15) = 1;
[Default: 1]
.TP
\fBxtpar2=\fP
Length scale for Plummer model (KZ(15) = 1)
[Default: 2]
.TP
\fBzmgas=\fP
Mass scale for logarithmic potential (KZ(15) = 2)
.TP
\fBrgas=\fP
Length scale for logarithmic potential (KZ(15) = 2)
.TP
\fBalphas=\fP
Power-law index for initial mass function (routine DATA)
[Default: 2.3]
.TP
\fBbody1=\fP
Maximum particle mass before scaling   
[Default: 5.0]
.TP
\fBbodyn=\fP
Minimum particle mass before scaling   
[Default: 1.0]
.TP
\fBq=\fP
Virial ratio (q=0.5 for virial equilibrium)  
[Default: 0]
.TP
\fBvxrot=\fP
XY-velocity scaling factor (> 0 for solid-body rotation)
[Default: 0]
.TP
\fBvzrot=\fP
Z-velocity scaling factor (not used if VXROT =
[Default: 0]
.TP
\fBrbar=\fP
Virial radius in pc (for scaling to physical
[Default: 1]
.TP
\fBzmbar=\fP
Mean mass in solar units   
[Default: 1]
.TP
\fBxcm=\fP
Displacement for subsystem (routine SCALE; KZ(17)) 
.TP
\fBecc=\fP
Eccentricity of relative motion for subsystem (ECC =< 1)
.SH EXAMPLES
\fBnbody2\fP can also be used to generate snapshots. Here is an example
to create a 1024 body homogeneous sphere in the directory
run1 and a plummer sphere in run2.
.nf
    runbody2 "" run1 1024 tcrit=0 kz=0,0,1
    u3tos run1/OUT3 run1/run1.snap
    runbody2 "" run2 1024 tcrit=0 kz=0,0,1,0,1
    u3tos run2/OUT3 run2/run2.snap
.fi
.PP
You can also supply existing NEMO snapshots as initial conditions
.nf
    runbody2 run4.snap run4 tcrit=0 kz=0,0,1,2
        (this option doesn't work yet)
.fi
.SH BUGS
Scaling can cause output to become out of bounds.
.SH SEE ALSO
nbody2(1NEMO), nbody0(1NEMO), runbody1(1NEMO), snapshot(5NEMO), u3tos(1NEMO), stou4(1NEMO)
.SH FILES
.nf
.ta +2i
$NEMO/src/nbody/evolve/aarseth/tools	code
$outdir/fort.1                       	restart dump (compile time dep. size)
$outdir/fort.2                         	restart dump (compile time dep. size)
$outdir/fort.4                         	restart dump (compile time dep. size)
$outdir/OUT3                        	particle dump (see \fIu3tos(1NEMO)\fP)
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
15-jun-97	V1.0 Created	PJT
15-jul-97	V1.2 out= now outdir=, nbody= now required value	PJT
5-mar-98	readying for new release	PJT
17-mar-06	V1.4 using fullname for in=	PJT
9-feb-2019	V2.0 better defaults for snapshot input	PJT
.fi


src/nbody/evolve/aarseth/tools/runbody2.c
/*
 *  MAIN.C: NEMO driver for nbody2
 *	
 *	16-jun-97   V1.0    does the basics, nothing fancy      PJT
 *      18-jun-97   V1.1    more logic builtin (see input.f, data.f, scale.f)
 *	21-jun-97   V1.1a   added history
 *	15-jul-97   V1.2    changed out= to outdir=, no def for nbody=     PJT
 *                          added tnext=
 *	 5-mar-98   V1.3    added KZ_ parameters in the code		   PJT
 *      17-mar-06   V1.4    using fullname() for in=                       pjt
 *      17-sep-2013 V1.5    using new run interface                        PJT
 *       9-feb-2017 V2.0    better defaults for snapshot input             PJT
 *
 */

#include <stdinc.h>
#include <getparam.h>
#include <filefn.h>
#include <run.h>

#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>

string defv[] = {
    "in=\n            input file (optional - see nbody= ) ",
    "outdir=???\n     output run directory (required)",

    "nbody=???\n      Total particle number (<= NMAX).",
    "nfix=1\n         Output frequency of data save or binaries; KZ(3 & 6)",
    "nrand=0\n        Random number sequence skip",
    "nnbmax=40\n      Maximum number of neighbours (< LMAX)",
    "nrun=1\n         Run identification index",

    "etai=0.02\n      Time-step parameter for irregular force polynomial",
    "etar=0.04\n      Time-step parameter for regular force polynomial",
    "rs0=1.0\n        Initial radius of neighbour sphere",
    "deltat=0.5\n     Output time interval in units of the crossing time",
    "tcrit=5.0\n      Termination time in units of the crossing time",
    "tnext=\n         Next output time for restart runs",
    "qe=0.00002\n     Energy tolerance (restart if DE/E > 5*QE & KZ(2) > 1)",
    "eps=0.01\n       Potential softening parameter",


    "kz=0,0,1,2,0,0,0,0,0,0,1,0,0,0,0,1,0,0,1,0\n  Non-zero options for alternative paths (see below)\n"
      " 1  COMMON save on unit 1 at end of run (=2: every 100*NMAX steps)\n"
      " 2  COMMON save on unit 2 at output (=1); restart if DE/E > 5*QE (=2)\n"
      " 3  Basic data written to unit 3 at output time (frequency NFIX)\n"
      " 4  Initial conditions on unit 4 (=1: output; =2: input)\n"
      " 5  Initial conditions (=0: uniform & isotropic; =1: Plummer model)\n"
      " 6  Output of significant binaries (=2: frequency NFIX)\n"
      " 7  Lagrangian radii (=1: unit 6; =2: unit 7; =3: both types)\n"
      " 8  Core radius & density centre (N > 20 only)\n"
      " 9  Individual bodies printed at output time (MIN(5**KZ9,N))\n"
     " 10  No unique density centre (skips velocity modification of RS(I))\n"
     " 11  Modification of ETAI & ETAR by tolerance QE\n"
     " 12  Inelastic mergers (>1: diagnostic output)\n"
     " 13  Escaper removal (R > 2*RTIDE; RTIDE = 10*RSCALE)\n"
     " 14  Skip full predictor loop if NNB > KZ(14) = <NNB> & KZ(14) > 0\n"
     " 15  External potential (=1: Plummer model; =2: logarithmic potential)\n"
     " 16  No scaling of initial conditions\n"
     " 17  Generation of two subsystems (merger experiment)\n"
     " 18  Adjustment of coordinates & velocities to c.m. condition\n"
     " 19  Use code units for tcrit/deltat",
     " 20  Not used at present"

    "xtpar1=\n        Mass of external Plummer model (KZ(15) = 1; scaled units)",
    "xtpar2=\n        Length scale for Plummer model (KZ(15) = 1)",
    "zmgas=\n         Mass scale for logarithmic potential (KZ(15) = 2)",
    "rgas=\n          Length scale for logarithmic potential (KZ(15) = 2)",

    "alphas=2.3\n     Power-law index for initial mass function",
    "body1=5.0\n      Maximum particle mass before scaling",
    "bodyn=1.0\n      Minimum particle mass before scaling",

    "q=0.0\n          Virial ratio (q=0.5 for virial equilibrium)",
    "vxrot=0.0\n      XY-velocity scaling factor (> 0 for solid-body rotation)",
    "vzrot=0.0\n      Z-velocity scaling factor (not used if VXROT = 0)",
    "rbar=1.0\n       Virial radius in pc (for scaling to physical units)",
    "zmbar=1.0\n      Mean mass in solar units",

    "xcm=\n           Displacement for subsystem (routine SCALE; KZ 17)",
    "ecc=\n           Eccentricity of relative motion for subsystem (ECC =< 1)",

    "kstart=1\n       Running mode (1=new 2=restart 3,4,5=restart w/ new par",
    "tcomp=40.0\n     Maximum allowed running time (minutes)",

    "VERSION=1.5\n    17-sep-2013 PJT",
    NULL,
};

string usage="Ahmad-Cohen N-body code";

#define KZ_MAX  20

#define KZ_OUT   3
#define KZ_INI   4
#define KZ_EXT  15
#define KZ_MER  17
#define KZ_COM  18

nemo_main()
{
    int nbody, nfix, nrand, nnbmax, nrun, kstart;
    real etai, etar, rs0, deltat, tcrit, qe, eps, tcomp;
    int k, nkz, kz[KZ_MAX];
    real alphas, body1, bodyn;
    real q, vxrot, vzrot, rbar, zmbar;
    string exefile = "nbody2";
    string parfile = "nbody2.in";
    string rundir = getparam("outdir");
    string infile, fname;
    char dname[256], runcmd[256];
    stream datstr, histr;

    kstart = getiparam("kstart");
    tcomp =  getdparam("tcomp");

    nbody = getiparam("nbody");
    nfix = getiparam("nfix");
    nrand = getiparam("nrand");
    nnbmax = getiparam("nnbmax");
    nrun = getiparam("nrun");

    etai = getdparam("etai");
    etar = getdparam("etar");
    rs0 = getdparam("rs0");
    deltat = getdparam("deltat");
    tcrit = getdparam("tcrit");
    qe = getdparam("qe");
    eps = getdparam("eps");

    nkz = nemoinpi(getparam("kz"),kz,20);
    if (nkz < 0) error("%d Syntax error kz=%s",nkz,getparam("kz"));
    for (k=nkz; k<KZ_MAX; k++) kz[k]=0;
    for (k=0; k<KZ_MAX; k++) dprintf(1,"%d ",kz[k]);
    dprintf(1,"\n");

    alphas = getdparam("alphas");
    body1 = getdparam("body1");
    bodyn = getdparam("bodyn");

    q = getdparam("q");
    vxrot = getdparam("vxrot");
    vzrot = getdparam("vzrot");
    rbar = getdparam("rbar");
    zmbar = getdparam("zmbar");

    infile = getparam("in");
    fname = fullname(infile);

    run_mkdir(rundir);

    sprintf(dname,"%s/%s",rundir,parfile);
    datstr = stropen(dname,"w");    

    /*  New Run */

  if (kstart == 1) {

    fprintf(datstr,"%d %g\n",kstart,tcomp);
    fprintf(datstr,"%d %d %d %d %d\n",nbody,nfix,nrand,nnbmax,nrun);
    fprintf(datstr,"%g %g %g %g %g %g %g\n",etai,etar,rs0,deltat,tcrit,qe,eps);

    for (k=0; k<KZ_MAX; k++) fprintf(datstr,"%d ",kz[k]);
    fprintf(datstr,"\n");

    if (kz[KZ_EXT-1] > 0) {           /* external potential parameters */
        if (kz[KZ_EXT-1]==1)
            fprintf(datstr,"%g %g\n",getdparam("xtpar1"),getdparam("xtpar2"));
        else
            fprintf(datstr,"%g %g\n",getdparam("zmgas"),getdparam("rgas"));
    }
            
    fprintf(datstr,"%g %g %g\n",alphas,body1,bodyn);
    fprintf(datstr,"%g %g %g %g %g\n",q,vxrot,vzrot,rbar,zmbar);

    if (kz[KZ_MER-1] > 0)
        fprintf(datstr,"%g %g\n",getdparam("xcm"), getdparam("ecc"));
    strclose(datstr);

    run_cd(rundir);
    histr = stropen("history","w");
    put_history(histr);
    strclose(histr);

    if (hasvalue("in")) {
	if (*infile == '-') {		/* do something special for pipes */
	  sprintf(runcmd,"stou4 %s",infile);
	} else {
	  sprintf(runcmd,"stou4 %s nbody=%d",fname,nbody);
        } 
        dprintf(0,"%s\n",runcmd);
        if (run_sh(runcmd)) error("Error converting input data");
    }

    sprintf(runcmd,"%s < %s",exefile,parfile);
    run_sh(runcmd);
  } else {
    error("kstart=%d not yet supported",kstart);
  }
}


